src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,745 @@
+    match Convert.texpr1_expr_of_cil_exp ask t t.env exp no_ov with
+    | texp -> assign_texpr t var texp
+    | exception Convert.Unsupported_CilExp _ ->
+      if is_bot_env t then t else forget_vars t [var]
+
+  let assign_exp ask t var exp no_ov =
+    let res = assign_exp ask t var exp no_ov in
+    if M.tracing then M.tracel "ops" "assign_exp t:\n %s \n var: %s \n exp: %a\n no_ov: %b -> \n %s\n"
+        (show t) (Var.to_string var) d_exp exp (Lazy.force no_ov) (show res) ;
+    res
+
+  let assign_var (t: VarManagement.t) v v' =
+    let t = add_vars t [v; v'] in
+    assign_texpr t v (Var v')
+
+  let assign_var t v v' =
+    let res = assign_var t v v' in
+    if M.tracing then M.tracel "ops" "assign_var t:\n %s \n v: %s \n v': %s\n -> %s\n" (show t) (Var.to_string v) (Var.to_string v') (show res) ;
+    res
+
+  (** Parallel assignment of variables.
+      First apply the assignments to temporary variables x' and y' to keep the old dependencies of x and y
+     and in a second round assign x' to x and y' to y
+  *)
+  let assign_var_parallel t vv's =
+    let assigned_vars = List.map fst vv's in
+    let t = add_vars t assigned_vars in
+    let primed_vars = List.init (List.length assigned_vars) (fun i -> Var.of_string (Int.to_string i  ^"'")) in (* TODO: we use primed vars in analysis, conflict? *)
+    let t_primed = add_vars t primed_vars in
+    let multi_t = List.fold_left2 (fun t' v_prime (_,v') -> assign_var t' v_prime v') t_primed primed_vars vv's in
+    match multi_t.d with
+    | Some arr when not @@ is_top multi_t ->
+      let switched_arr = List.fold_left2 (fun multi_t assigned_var primed_var-> assign_var multi_t assigned_var primed_var) multi_t assigned_vars primed_vars in
+      drop_vars switched_arr primed_vars ~del:true
+    | _ -> t
+
+  let assign_var_parallel t vv's =
+    let res = assign_var_parallel t vv's in
+    if M.tracing then M.tracel "ops" "assign_var parallel: %s -> %s \n" (show t) (show res);
+    res
+
+  let assign_var_parallel t vv's = timing_wrap "var_parallel" (assign_var_parallel t) vv's
+
+  let assign_var_parallel_with t vv's =
+    let t' = assign_var_parallel t vv's in
+    t.d <- t'.d;
+    t.env <- t'.env
+
+  let assign_var_parallel_with t vv's =
+    if M.tracing then M.tracel "var_parallel" "assign_var parallel'\n";
+    assign_var_parallel_with t vv's
+
+  let assign_var_parallel' t vs1 vs2 =
+    let vv's = List.combine vs1 vs2 in
+    assign_var_parallel t vv's
+
+  let assign_var_parallel' t vv's =
+    let res = assign_var_parallel' t vv's in
+    if M.tracing then M.tracel "ops" "assign_var parallel'\n";
+    res
+
+  (* This is supposed to the be the backwards transformer for assign, not sure this is correct *)
+  let substitute_exp ask t var exp no_ov =
+    let t = if not @@ Environment.mem_var t.env var then add_vars t [var] else t in
+    let res = assign_exp ask t var exp no_ov in
+    forget_vars res [var]
```

Comment:
Is this really correct? It looks somehow like an expensive `forget`?

Reply:
a grep on the src folder does not reveal any usages of this function outside of the apron universe. Are we sure, we actually need this?

Reply:
Apart from this, `assign_exp` essentially adds an equality to the domainvalue, and then forgets about `var`. It is more expensive then a regular `forget_vars` since it may leave behind some connections between `exp` and whatever links `var` had in the old domainvalue `t` .

Reply:
After some discussions we determined it is indeed correct.